appIcons.js

```diff
@@ -1740,60 +1740,74 @@ var MyShowAppsIconMenu = Utils.defineClass({
 
     _dtpRedisplay: function() {
         this.removeAll();
+        
+        // Only add menu entries for commands that exist in path
+        function _appendItem(obj, info) {
+            if (Utils.checkExists(info.cmd[0])) {
```

Comment:
I had a thought about this (sorry, should have been sooner), maybe we should validate that the commands exist beforehand and keep that information available so we don't have to call `which` for every command each time we are opening the context menu. Not sure if this would be better "performance/memory" wise. What do you think?

Reply:
Performance/wise it is certainly better. Memory wise, maybe not, but I have a suggestion on this.
I think there is an app settings store in dash-to-panel that stores config settings that can be configured by an administrator through gsettings. Is that right? Is there another store where I could store volatile state that gets reset everytime the extension is reloaded? (I'm loathe to use a global variable if there is a better pattern)

I'm working on removing the hardcoded menu entries from MyShowAppsIconMenu and storing them in the global settings as a list (kind of like an MRU list). I can then loop through each "MRU" item and add them to the menu. This way an admin (or someone building a distro) could customize the apps shown on the menu, and the added benefit of customizing the list would justify using more memory. Moreover, if someone builds a settings UI to customize the menu, they would already have a place to store the list into.

As I add each item, I could flag each app as available/missing the first time the menu is opened, and then never have to check again. If I use the volatile store to store the flags, the check will run once every time the extension is loaded.

What do you think?

Reply:
That sounds great! I wanted to do something like that, but the UI part of the settings got me to put it aside for now :) but nothing prevents us from doing the gsettings now and plug back the UI later on. 

I'm not aware of a volatile store (doesn't mean there isn't one), but honestly at that point a module variable containing key value pairs of the commands and bools stating if available or not sounds reasonable to me. Of course there is a chance that a user installs a missing command and that it won't be picked up until the next time the extension is loaded, but I don't think that is an issue here.

Thanks!

Reply:
Latest revision should address this. I'm saving the list of menu entries in four global settings:
show-apps-button-context-menu-commands
show-apps-button-context-menu-titles
panel-context-menu-commands
panel-context-menu-titles

I am also caching the check results in checkedCommandsMap. The first right click will incur the penalty of checking for each command, but clicking again will read from cache. I haven't seen a real perf improvement, but I haven't tried very slow hard drives yet.